server/app/auth/ApiAuthenticator.java

```diff
@@ -0,0 +1,97 @@
+package auth;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.time.Instant;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import models.ApiKey;
+import org.apache.commons.net.util.SubnetUtils;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.credentials.UsernamePasswordCredentials;
+import org.pac4j.core.credentials.authenticator.Authenticator;
+import org.pac4j.core.exception.BadCredentialsException;
+import play.cache.NamedCache;
+import play.cache.SyncCacheApi;
+import services.apikey.ApiKeyService;
+
+/**
+ * Authenticator for API requests based on HTTP basic auth and backed by the {@link ApiKey}
+ * resource. Background: https://en.wikipedia.org/wiki/Basic_access_authentication
+ *
+ * <p>When referenced by a request, {@code ApiKey}s are stored in the "api-keys" named cache, keyed
+ * by their key ID. This allows controller code to load the key without the need for a subsequent
+ * database query. This is necessary because pac4j creates profiles in the next step, but we need
+ * the API key to perform authentication (this step).
+ *
+ * <p>Note that at this layer, the request is authenticated, not authorized. All API requests that
+ * reach a controller are already authenticated, but it is the controller's responsibility to check
+ * that the resource being accessed is authorized for the request's API key.
+ *
+ * <p>The API key ID and secret are provided as the basic auth username and password, respectively.
+ * To be valid, a request must reference a valid API key with its username credential, and further
+ * that key must:
+ * <li>Not be retired.
+ * <li>Have an expiration date in the future.
+ * <li>Have a subnet that includes the IP address of the request.
+ * <li>Have a salted key secret that matches the salted password in the request's basic auth
+ *     credentials.
+ */
+public class ApiAuthenticator implements Authenticator {
+
+  private static final int CACHE_EXPIRATION_TIME_SECONDS = 60;
+
+  private final SyncCacheApi syncCacheApi;
+  private final Provider<ApiKeyService> apiKeyService;
+
+  @Inject
+  public ApiAuthenticator(
+      @NamedCache("api-keys") SyncCacheApi syncCacheApi, Provider<ApiKeyService> apiKeyService) {
+    this.syncCacheApi = checkNotNull(syncCacheApi);
+    this.apiKeyService = checkNotNull(apiKeyService);
+  }
+
+  @Override
+  public void validate(Credentials rawCredentials, WebContext context, SessionStore sessionStore) {
+    if (!(rawCredentials instanceof UsernamePasswordCredentials)) {
+      throw new RuntimeException("ApiAuthenticator must receive UsernamePasswordCredentials.");
+    }
+
+    UsernamePasswordCredentials credentials = (UsernamePasswordCredentials) rawCredentials;
+    String keyId = credentials.getUsername();
```

Comment:
Are we overloading credentials in a confusing way by making a username and keyId synonymous?

Is there a sketch of this in your design doc?

Reply:
The design specifies that we're using basic auth to authenticate API requests. "username" and "password" is the parlance basic auth uses for describing those values. Our internal consumption of them as "key ID" and "key secret" matches the CiviForm concepts, rather than the protocol ones.

Reply:
re: the slack convo, can you doc this here so it's clearer why this is happening and that it is ok despite looking odd.

Reply:
added a comment